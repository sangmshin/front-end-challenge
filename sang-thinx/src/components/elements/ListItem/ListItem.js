import React from 'react';
import { func, string } from 'prop-types';
import cx from 'classnames';
import { Button } from 'components/elements';
import styles from './ListItem.module.scss';

export const ListItem = ({
  item,
  onSelect,
  selectedItem,
}) => {
  const isSelected = selectedItem === item;
  const handleClick = () => onSelect(item);
  const buttonClass = cx({
    [styles.item]: true,
    [styles.selected]: isSelected
  });

  return (
    <li role='option' aria-selected={isSelected}>
      <Button
        ariaLabel={`size: ${item}`}
        className={buttonClass}
        onClick={handleClick}
        role='option radio'
      >
        {item}
      </Button>
    </li>
  )
}

ListItem.propTypes = {
  item: string.isRequired,
  onSelect: func.isRequired,
  selectedItem: string,
};

ListItem.defaultProps = {
  selectedItem: null,
};

export default ListItem;