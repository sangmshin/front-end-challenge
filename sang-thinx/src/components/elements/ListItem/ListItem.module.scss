@import 'variables';

.item {
  width: 100%;
  text-align: left;
  padding: 13px 17px;

  &:hover {
    background-color: $primary-beige;
  }
}

.selected {
  background-color: $primary-beige;
}