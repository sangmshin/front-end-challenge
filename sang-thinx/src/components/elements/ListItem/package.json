{
  "name": "ListItem",
  "main": "./ListItem.js"
}
