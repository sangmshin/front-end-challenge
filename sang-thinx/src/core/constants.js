export const COLOR_HEXCODE_MAP = {
  beige: '#f2d9a5',
  black: '#000',
}