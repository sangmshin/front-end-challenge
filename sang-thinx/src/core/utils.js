const SMALL = 768; /* Small screen / tablet */
const MEDIUM = 1024; /* Medium screen / desktop */
const LARGE = 1200; /* Large screen / wide desktop */
let BREAKPOINT;

/**
 * Finds the current matched media query breakpoint
 *
 * @returns {String} matched media query breakpoint [large, medium, small, xsmall]
 */
export function getCurrentBreakPoint() {
  if (!BREAKPOINT) {
    BREAKPOINT = {
      large: window.matchMedia(`(min-width: ${LARGE}px)`),
      medium: window.matchMedia(`(min-width: ${MEDIUM}px)`),
      small: window.matchMedia(`(min-width: ${SMALL}px)`),
      xsmall: window.matchMedia(`(max-width: ${SMALL - 1}px)`)
    };
  }

  return Object.keys(BREAKPOINT).find(breakpoint => BREAKPOINT[breakpoint].matches);
}

/**
 * Determines if the data is array and has at least one or more items
 * @return {boolean}
 */
export const isArrayWithLength = arr => Array.isArray(arr) && arr.length > 0;

/**
 * A function with no operation. (for testing)
 */
export const noop = () => null;
