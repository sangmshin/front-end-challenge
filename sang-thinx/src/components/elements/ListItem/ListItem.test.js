import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { ListItem } from './ListItem';
import { noop } from 'core/utils';

const props = {
  item: 'Banana',
  onSelect: noop,
  selectedItem: 'Strawberry'
}

describe('Component :: ListItem', () => {
  describe('#render', () => {
    const component = shallow(
      <ListItem {...props} />
    );
    it('should match snapshot', () => {
      expect(toJson(component)).toMatchSnapshot();
    });
  });

  describe('#selectedItem', () => {
    const component = shallow(
      <ListItem {...props} selectedItem='Banana' />
    );
    it('should attach "selected" style when "selectedItem" prop match "item" prop', () => {
      expect(component.find('ForwardRef').prop('className')).toBe('item selected');
    });
  });

  describe('#click in the option', () => {
    const spy = jest.fn();
    const component = shallow(
      <ListItem {...props} onSelect={spy} />
    );
    it('should executes the onSelect', () => {
      component.find('ForwardRef').simulate('click');
      expect(spy).toHaveBeenCalled();
    });
  });
});
