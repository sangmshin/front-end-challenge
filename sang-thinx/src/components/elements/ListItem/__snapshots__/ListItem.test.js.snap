// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component :: ListItem #render should match snapshot 1`] = `
<li
  aria-selected={false}
  role="option"
>
  <ForwardRef
    accessibleText=""
    ariaLabel="size: Banana"
    className="item"
    disabled={false}
    onClick={[Function]}
    role="option radio"
    style={null}
  >
    Banana
  </ForwardRef>
</li>
`;
